pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,259 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	v1authenticationapi "k8s.io/api/authentication/v1"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	apiserverserviceaccount "k8s.io/apiserver/pkg/authentication/serviceaccount"
+	clientset "k8s.io/client-go/kubernetes"
+	v1authentication "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+
+	"k8s.io/klog"
+)
+
+// defaultExpiration defines the duration of a TokenRequest in seconds.
+var defaultExpiration = int64(3600)
+
+type configSlot struct {
+	config  *restclient.Config
+	expTime metav1.Time
+}
+
+type DynamicControllerClientBuilder struct {
+	// ClientConfig is a skeleton config to clone and use as the basis for each controller client
+	ClientConfig *restclient.Config
+
+	// CoreClient is used to provision service accounts if needed and watch for their associated tokens
+	// to construct a controller client
+	CoreClient v1core.CoreV1Interface
+
+	// AuthenticationClient is used to check API tokens to make sure they are valid before
+	// building a controller client from them
+	AuthenticationClient v1authentication.AuthenticationV1Interface
+
+	// Namespace is the namespace used to host the service accounts that will back the
+	// controllers.  It must be highly privileged namespace which normal users cannot inspect.
+	Namespace string
+
+	// configs is a concurrent-safe Map which stores all the configs
+	configs sync.Map
+
+	clock clock.Clock
+}
+
+func NewDynamicBuilder(clientConfig *restclient.Config, coreClient v1core.CoreV1Interface, authClient v1authentication.AuthenticationV1Interface, ns string) ControllerClientBuilder {
+	builder := &DynamicControllerClientBuilder{
+		ClientConfig:         clientConfig,
+		CoreClient:           coreClient,
+		AuthenticationClient: authClient,
+		Namespace:            ns,
+		configs:              sync.Map{},
+		clock:                clock.RealClock{},
+	}
+	builder.startSync()
+	return builder
+}
+
+func (t *DynamicControllerClientBuilder) Config(name string) (*restclient.Config, error) {
+	sa, err := t.getOrCreateServiceAccount(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if i, ok := t.configs.Load(name); ok {
+		return i.(*configSlot).config, nil
+	}
+
+	username := apiserverserviceaccount.MakeUsername(sa.Namespace, sa.Name)
+	clientConfig := restclient.AnonymousClientConfig(t.ClientConfig)
+	clientConfig.BearerToken = ""
```

Comment:
clearing the token is not necessary... anonymousclientconfig returns a config with no bearer token